### PR TITLE
WP3: apply clang-tidy --checks=modernize-return-braced-init-list --fix

### PR DESCRIPTION
**Pull request by aphecetche**
as agreed in WP3 meeting, here are the fixes for the `modernize-return-braced-init-list` warnings from `clang-tidy`. This PR is to see if nothing breaks.

### CODE REVIEW

**Comment by davidrohr** [comment]
FYI: You have to revert this line to reenable the check in the codechecker: https://github.com/alisw/alidist/pull/3506/files#diff-35397bd99ec0ce432c089b4962d848df6faaf20441ceaa66d44e344b7f61bc0dL56

**Comment by ktf** [comment]
Isn't this change actually dangerous? Take:

```C++
struct Foo {
    int i;
};

struct Bar {
    float f = 0;
    int i = 0;
};

Foo foo() {
    return {1};
}

int foobar() {
    return foo().i;
}
```

if I change foo() to return Bar, it will still compile (because int casts to float automatically) but most likely it will not do what one wants. On the other hand, the version where one has:

```
Foo foo() {
  return Foo{1};
}
```

will complain that Foo cannot be converted to Bar. In this particular case, it's fine because the function is trivial enough. However, for large enough methods with multiple return statements, it suddenly becomes possible to miss a substitution and still to be able to compile.

File: run-clang-tidy [hunk 1]
```diff
@@ -0,0 +1,347 @@
+#!/usr/bin/env python
+#
+#===- run-clang-tidy.py - Parallel clang-tidy runner --------*- python -*--===#
```
Comment (sawenzel) [inline run-clang-tidy]: this file should probably not be committed

File: run_clang_tidy.log [hunk 2]
```diff
@@ -0,0 +1,2832 @@
+Enabled checks:
+    modernize-return-braced-init-list
```
Comment (sawenzel) [inline run_clang_tidy.log]: this file should probably not be committed

Comment (aphecetche) [reply run_clang_tidy.log]: @sawenzel indeed. will remove that one and the one above when updating the PR if / when we decide to go with (and_ when the fair logger related PRs are done, as they created some conflict)

File: Detectors/MUON/MCH/Raw/ElecMap/src/FeeLinkId.cxx [hunk 3]
```diff
@@ -31,7 +31,7 @@ FeeLinkId decodeFeeLinkId(uint32_t x)
 {
   uint16_t feeId = static_cast<uint16_t>((x & 0xFFFF0000) >> 16);
   uint16_t linkId = static_cast<uint8_t>(x & 0xF);
-  return FeeLinkId(feeId, linkId);
+  return {feeId, linkId};
```
Comment (ihrivnac) [inline Detectors/MUON/MCH/Raw/ElecMap/src/FeeLinkId.cxx:34]: This code produces a compiler error on Mac builder:

```
/System/Volumes/Data/build/ci_checks/alice_ci_0/o2/sw/SOURCES/O2/7714/0/Detectors/MUON/MCH/Raw/ElecMap/src/FeeLinkId.cxx:34:18: error: non-constant-expression cannot be narrowed from type 'uint16_t' (aka 'unsigned short') to 'uint8_t' (aka 'unsigned char') in initializer list [-Wc++11-narrowing]
  return {feeId, linkId};
                 ^~~~~~
```
The` linkId` is declared as `uint16_t` while it takes value of  `uint8_t`, what is also expected in the class constructor.

**Comment by ktf** [comment]
q.e.d. let’s just forget about that option…

Ciao,
Giulio

**Comment by ihrivnac** [comment]
> q.e.d. let’s just forget about that option… Ciao, Giulio
> […](#)
> 

The compiler error happens due to implicit narrowing at `FeeLinkId(feeId, linkId)`,  which passes silently with the current code, so it speaks rather for accepting the option, and fixing the code where needed; in this case

`uint8_t linkId = static_cast<uint8_t>(x & 0xF);
`

Or using `auto` would also work here.

**Comment by ktf** [comment]
The problem I have is that the suggested fix is to remove the type, not to add auto or explicit, possibly poisoning our code of hidden bugs which might take long to understand. All this to end up with code which will look almost exactly like the original one.

**Comment by ihrivnac** [comment]
> 

The example above shows just the opposite: it reveals a hidden bug in the code.
Returning via the class construction may induce an implicit conversion, while when the type is omitted, it is guaranteed that the exact type will be constructed from the initialization list.

**Comment by ktf** [comment]
This has now conflicts, could you regenerate it?

**Review by ktf — COMMENTED** [review]
This has now conflicts, could you regenerate it?

**Comment by ihrivnac** [comment]
@aphecetche , 
There is another narrowing arror (MacOS build):

```
/System/Volumes/Data/build/ci_checks/alice_ci_0/o2/sw/SOURCES/O2/7714/0/Framework/GUISupport/src/FrameworkGUIDebugger.cxx:269:21: error: non-constant-expression cannot be narrowed from type 'size_t' (aka 'unsigned long') to 'double' in initializer list [-Wc++11-narrowing]
            return {histoData->time[pos], ((int*)(histoData->points))[pos]};
                    ^~~~~~~~~~~~~~~~~~~~

```
Can you commit a fix by changing
`            return {histoData->time[pos], ((int*)(histoData->points))[pos]};
`
with
`            return {static_cast<double>(histoData->time[pos]), ((int*)(histoData->points))[pos]};
`

And similarly for the other errors of this type, if they happen after this fix?

**Comment by aphecetche** [comment]
@ihrivnac will do as soon as I get a chance

**Comment by aphecetche** [comment]
@ihrivnac with the relevant modifications this one is now green.

**Comment by ihrivnac** [comment]
@aphecetche , thank you. I will add the link to the changes in this PR in the discussion item in the WP3 online notes for the discussion tomorrow.

**Pull request closed without merging**